modules/core/src/lifecycle/props.js

```diff
@@ -73,41 +78,31 @@ export function compareProps({
     return `${triggerName} changed shallowly`;
   }
 
-  // Test if new props different from old props
-  for (const key in oldProps) {
+  // Compare explicitly defined new props against old/default values
+  for (const key of Object.keys(newProps)) {
     if (!(key in ignoreProps)) {
-      if (!(key in newProps)) {
-        return `${triggerName}.${key} dropped`;
-      }
-      const newProp = newProps[key];
-      const oldProp = oldProps[key];
-      const propType = propTypes[key];
-
-      // If prop type has an equal function, invoke it
-      let equal = propType && propType.equal;
-      if (equal && !equal(newProp, oldProp, propType)) {
-        return `${triggerName}.${key} changed deeply`;
-      }
-
-      if (!equal) {
-        // If object has an equals function, invoke it
-        equal = newProp && oldProp && newProp.equals;
-        if (equal && !equal.call(newProp, oldProp)) {
-          return `${triggerName}.${key} changed deeply`;
-        }
+      if (!(key in oldProps)) {
+        return `${triggerName}.${key} added`;
       }
-
-      if (!equal && oldProp !== newProp) {
-        return `${triggerName}.${key} changed shallowly`;
+      const changed = comparePropValues(newProps[key], oldProps[key], propTypes[key]);
+      if (changed) {
+        return `${triggerName}.${key} ${changed}`;
       }
     }
   }
 
-  // Test if any new props have been added
-  for (const key in newProps) {
+  // Test if any old props have been dropped
+  for (const key of Object.keys(oldProps)) {
     if (!(key in ignoreProps)) {
-      if (!(key in oldProps)) {
-        return `${triggerName}.${key} added: undefined -> ${newProps[key]}`;
+      if (!(key in newProps)) {
+        return `${triggerName}.${key} dropped`;
+      }
+      if (!Object.hasOwnProperty.call(newProps, key)) {
```

Comment:
Just curious, what does this do? Does it compare further up the prototype chain?

Reply:
The first for loop already checked all the own properties of `newProps`, and in this second for loop we only want to check props that are in `oldProps` but not `newProps`.